pycheribuild/projects/adduser.py

```diff
@@ -0,0 +1,91 @@
+# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
+# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+# OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+# HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+# OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+# SUCH DAMAGE.
+#
+
+import os
+import getpass
+import subprocess
+import sys
+
+from pathlib import Path
+
+from .project import CheriConfig, SimpleProject, DefaultInstallDir
+from ..utils import (AnsiColour, coloured, status_update)
+from ..processutils import run_command
+
+
+class AddUser(SimpleProject):
+    target = "adduser"
+    native_install_dir = DefaultInstallDir.BOOTSTRAP_TOOLS
+
+    def __init__(self, config: CheriConfig):
+        super().__init__(config)
+
+    def process(self):
+        file = str(self.config.output_root.absolute()) + "/Dockerfile.adduser"
+        try:
+            user = getpass.getuser()
+        except KeyError:
+            user = "nobody"
+
+        # Create a Dockerfile that will contain this user's name, gid, uid
+        try:
+            fp = open(file, mode="w")
+            contents = "FROM cheribuild-docker\n\nRUN addgroup --gid " + \
+                       str(os.getgid()) + " " + user + \
+                       " && adduser --uid " + str(os.getuid()) + " --ingroup " + \
+                       user + " " + user + " \n"
+
+            fp.write(contents)
+
+            fp.close()
+
+        except OSError:
+            status_update("Could not create ", file)
+            sys.exit()
+
+        # Build a new image from our installed image with this user
+        try:
+            docker_run_cmd = ["docker", "build", "--tag=cheribuild-docker",
+                              "-f", file, "."]
+            run_command(docker_run_cmd, config=self.config)
+
+            os.remove(file)
+
+        except subprocess.CalledProcessError as e:
+            os.remove(file)
+
+            # if the image is missing print a helpful error message:
+            if e.returncode == 125:
+                cheribuild_dir = str(Path(__file__).absolute().parent.parent)
+                status_update("It seems like the docker image", self.config.docker_container, "was not found.")
+                status_update("In order to build the default docker image for cheribuild (cheribuild-docker) run:")
+                print(coloured(AnsiColour.blue, "cd", cheribuild_dir +
+                               "/docker && docker build --tag cheribuild-docker ."))
+                sys.exit(coloured(AnsiColour.red, "Failed to start docker!"))
```

Comment:
Old comment still applies, just now hidden as you changed the whitespace:

> This makes no sense. Use `self.fatal` with `fixit_hint=...`

Reply:
Funny, this is exactly the code in main that spits out the docker message.

Reply:
I stand corrected somewhat, apparently sys.exit, if given anything other than an int, will print that out and exit with code 1. Ew. But my point of self.fatal still stands; code in __main__.py is running outside the context of a target, so the interfaces, and expected behaviours, are very different (no project to call methods on, and if there's an error then it can only call sys.exit, whereas inside a target the expectation is that only the target fails).